flax/optim/adafactor.py

```diff
@@ -0,0 +1,166 @@
+        LR is markedly different from Adam, one doesn't use the 1/sqrt(hidden)
+        correction for this optimizer with attention-based models.
+      factored: boolean: whether to use factored second-moment estimator for 2d
+        variables.
+      multiply_by_parameter_scale: boolean: if True, then scale provided
+        learning_rate by parameter norm. if False, provided learning_rate is
+        absolute step size.
+      do_clipping: whether to clip gradients; if True, set clipping_threshold.
+      do_momentum: whether to use momentum; if True, set beta1.
+      beta1: a float value between 0 and 1, enables momentum and uses extra
+        memory if nonzero!  Off by default.
+      decay_rate: float: controls second-moment exponential decay schedule.
+      clipping_threshold: an optional float >= 1, if None no update clipping.
+      weight_decay_rate: rate at which to decay weights.
+      epsilon1: Regularization constant for squared gradient.
+      epsilon2: Regularization constant for parameter scale.
+    """
+    hyper_params = _AdafactorHyperParams(
+        learning_rate, factored, multiply_by_parameter_scale,
+        do_clipping, do_momentum, beta1, decay_rate, clipping_threshold,
+        weight_decay_rate, epsilon1, epsilon2)
+    super().__init__(hyper_params)
+
+  @staticmethod
+  def _decay_rate_pow(i, exponent=0.8):
+    """Default Adafactor second-moment decay schedule."""
+    t = jnp.array(i, jnp.float32) + 1.0
+    return 1.0 - t**(-exponent)
+
+  def init_param_state(self, param):
+    shape = param.shape
+    state = {k: jnp.zeros(()) for k in ['v_row', 'v_col', 'v', 'm']}
+    if self.hyper_params.factored and len(shape) >= 2:
+      state['v_row'] = jnp.zeros(shape[:-1], dtype=jnp.float32)
+      state['v_col'] = jnp.zeros(shape[:-2] + shape[-1:], dtype=jnp.float32)
+    else:
+      state['v'] = jnp.zeros_like(param)
+    if self.hyper_params.do_momentum:
+      state['m'] = jnp.zeros_like(param)
+    return _AdafactorParamState(**state)
+
+  def apply_param_gradient(self, step, hyper_params, param, state, grad):
+    assert hyper_params.learning_rate is not None, 'no learning rate provided.'
+    learning_rate = hyper_params.learning_rate
+    beta1 = hyper_params.beta1
+    decay_rate = hyper_params.decay_rate
+    clipping_threshold = hyper_params.clipping_threshold
+    weight_decay_rate = hyper_params.weight_decay_rate
+    epsilon1 = hyper_params.epsilon1
+    epsilon2 = hyper_params.epsilon2
+
+    updates = {k: jnp.zeros(()) for k in ['v_row', 'v_col', 'v', 'm']}
+    decay_rate = self._decay_rate_pow(step, exponent=decay_rate)
+    update_scale = learning_rate
+    if self.hyper_params.multiply_by_parameter_scale:
+      update_scale *= jnp.maximum(
+          jnp.sqrt(jnp.mean(param * param)), epsilon2)
+    mixing_rate = 1.0 - decay_rate
+
+    grad_sqr = grad * grad + epsilon1
+    if self.hyper_params.factored and len(param.shape) >= 2:
```

Comment:
if I understand correctly factorization is only performant on the last 2 dims. Perhaps that's good to mention explicitly.